File: 3510.minimum-pair-removal-to-sort-array-ii.py
#
# @lc app=leetcode id=3510 lang=python3
#
# [3510] Minimum Pair Removal to Sort Array II
#

# @lc code=start
import heapq
from typing import List

class Node:
    def __init__(self, val: int, min_idx: int):
        self.val = val
        self.min_idx = min_idx
        self.id = 0
        self.prev = None
        self.next = None

# @lc code=end
class Solution:
    def minimumPairRemoval(self, nums: List[int]) -> int:
        n = len(nums)
        if n <= 1:
            return 0
        
        nodes = []
        id_counter = 0
        id_to_node = {}
        for i, val in enumerate(nums):
            node = Node(val, i)
            node.id = id_counter
            id_counter += 1
            id_to_node[node.id] = node
            nodes.append(node)
        
        for i in range(n - 1):
            nodes[i].next = nodes[i + 1]
            nodes[i + 1].prev = nodes[i]
        
        # Compute initial violations
        violations = 0
        cur = nodes[0]
        while cur.next:
            if cur.val > cur.next.val:
                violations += 1
            cur = cur.next
        
        if violations == 0:
            return 0
        
        # Initialize priority queue
        pq = []
        cur = nodes[0]
        while cur.next:
            s = cur.val + cur.next.val
            heapq.heappush(pq, (s, cur.min_idx, cur.id, cur.next.id))
            cur = cur.next
        
        ops = 0
        while violations > 0:
            # Find valid minimum pair
            while pq:
                s, minidx, lid, rid = heapq.heappop(pq)
                if lid not in id_to_node or rid not in id_to_node:
                    continue
                node_l = id_to_node[lid]
                node_r = id_to_node[rid]
                if node_l.next != node_r:
                    continue
                # Valid pair found
                break
            else:
                # No more pairs, should not happen if violations > 0
                return ops
            
            # Update violations delta
            delta = 0
            p = node_l.prev
            if p:
                delta -= 1 if p.val > node_l.val else 0
            delta -= 1 if node_l.val > node_r.val else 0
            nxt = node_r.next
            if nxt:
                delta -= 1 if node_r.val > nxt.val else 0
            
            new_val = node_l.val + node_r.val
            if p:
                delta += 1 if p.val > new_val else 0
            if nxt:
                delta += 1 if new_val > nxt.val else 0
            
            violations += delta
            
            # Create new node
            new_node = Node(new_val, min(node_l.min_idx, node_r.min_idx))
            new_node.id = id_counter
            id_counter += 1
            id_to_node[new_node.id] = new_node
            
            # Relink
            new_node.prev = p
            new_node.next = nxt
            if p:
                p.next = new_node
            if nxt:
                nxt.prev = new_node
            
            # Deactivate old nodes
            del id_to_node[lid]
            del id_to_node[rid]
            
            # Push new pairs
            if p:
                news = p.val + new_node.val
                heapq.heappush(pq, (news, p.min_idx, p.id, new_node.id))
            if nxt:
                news = new_node.val + nxt.val
                heapq.heappush(pq, (news, new_node.min_idx, new_node.id, nxt.id))
            
            ops += 1
        
        return ops

File: 3510.minimum-pair-removal-to-sort-array-ii.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * @lc app=leetcode id=3510 lang=cpp
 *
 * [3510] Minimum Pair Removal to Sort Array II
 */

// @lc code=start
class Solution {
public:
    struct Node {
        long long val;
        int prev = -1, next = -1;
        int pos = 0;      // leftmost original index of this segment
        int ver = 0;      // increments when val changes
        bool alive = true;
    };

    struct Entry {
        long long sum;
        int pos;
        int left, lver;
        int right, rver;
    };

    struct Cmp {
        bool operator()(const Entry& a, const Entry& b) const {
            if (a.sum != b.sum) return a.sum > b.sum;   // min-heap by sum
            return a.pos > b.pos;                       // tie: leftmost
        }
    };

    int minimumPairRemoval(vector<int>& nums) {
        int n = (int)nums.size();
        if (n <= 1) return 0;

        vector<Node> nodes(n);
        for (int i = 0; i < n; i++) {
            nodes[i].val = (long long)nums[i];
            nodes[i].pos = i;
            nodes[i].prev = i - 1;
            nodes[i].next = (i + 1 < n ? i + 1 : -1);
        }

        long long inversions = 0;
        for (int i = 0; i + 1 < n; i++) {
            if ((long long)nums[i] > (long long)nums[i + 1]) inversions++;
        }
        if (inversions == 0) return 0;

        priority_queue<Entry, vector<Entry>, Cmp> pq;

        auto pushPair = [&](int left) {
            if (left == -1 || !nodes[left].alive) return;
            int right = nodes[left].next;
            if (right == -1 || !nodes[right].alive) return;
            pq.push(Entry{
                nodes[left].val + nodes[right].val,
                nodes[left].pos,
                left, nodes[left].ver,
                right, nodes[right].ver
            });
        };

        for (int i = 0; i + 1 < n; i++) pushPair(i);

        auto isInv = [&](int a, int b) -> bool {
            if (a == -1 || b == -1) return false;
            // a and b are expected alive when called
            return nodes[a].val > nodes[b].val;
        };

        int ops = 0;
        while (true) {
            // get valid minimum pair
            Entry e;
            while (true) {
                e = pq.top();
                pq.pop();
                int l = e.left, r = e.right;
                if (l == -1 || r == -1) continue;
                if (!nodes[l].alive || !nodes[r].alive) continue;
                if (nodes[l].ver != e.lver || nodes[r].ver != e.rver) continue;
                if (nodes[l].next != r) continue;
                // valid
                break;
            }

            int x = e.left;
            int y = e.right;           // y == nodes[x].next
            int p = nodes[x].prev;
            int nn = nodes[y].next;

            // remove old inversion contributions
            if (p != -1 && isInv(p, x)) inversions--;
            if (isInv(x, y)) inversions--;
            if (nn != -1 && isInv(y, nn)) inversions--;

            // merge y into x
            nodes[x].val += nodes[y].val;
            nodes[x].ver++;
            nodes[x].next = nn;
            if (nn != -1) nodes[nn].prev = x;

            nodes[y].alive = false;
            nodes[y].prev = nodes[y].next = -1;

            // add new inversion contributions
            if (p != -1 && isInv(p, x)) inversions++;
            if (nn != -1 && isInv(x, nn)) inversions++;

            // update affected pairs in heap
            if (p != -1) pushPair(p);
            pushPair(x);

            ops++;
            if (inversions == 0) return ops;
        }
    }
};
// @lc code=end
